core/src/raw/oio/read/buffer_reader.rs

```diff
@@ -0,0 +1,457 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use bytes::BufMut;
+use bytes::{Bytes, BytesMut};
+
+use tokio::io::ReadBuf;
+
+use std::cmp::min;
+use std::io::SeekFrom;
+
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+
+use crate::raw::*;
+use crate::*;
+
+/// [BufferReader] allows the underlying reader to fetch data at the buffer's size
+/// and is used to amortize the IO's overhead.
+pub struct BufferReader<R> {
+    r: R,
+    cur: u64,
```

Comment:
Maybe we don't need to track the `cur` of underlying read? We just need a `pos` like [BufReader](https://docs.rs/tokio/latest/src/tokio/io/util/buf_reader.rs.html).

- If `pos == 0`, we can forward the seek to underlying directly. 
- If `pos > 0`, we can calculate the correct `SeekFrom` for seek.

Reply:
> Maybe we don't need to track the `cur` of underlying read? We just need a `pos` like [BufReader](https://docs.rs/tokio/latest/src/tokio/io/util/buf_reader.rs.html).
> 
> * If `pos == 0`, we can forward the seek to underlying directly.
> * If `pos > 0`, we can calculate the correct `SeekFrom` for seek.

I’m going to check this.